### PR TITLE
[FSDP][Replicate] added replicate function that uses FSDP instead of DDP

### PR DESCRIPTION
**Pull request by anshul-si**
**Summary**
Users would like to use Replicate with TP. Currently, the replicate function uses DDP, which has not been maintained resulting in a lack of integration options. Since users can use FSDP with TP, we will make the replicate function use FSDP so that users can use replicate with FSDP. To that end I have created a replicate function that uses FSDP instead of DDP. One blocker that I ran into is that the replicate function has a contract which assigns a module "replicate" attribute in registry. This would mean that fully_shards is_composable requirement would not be satisfied making it impossible to apply fully_shard to a replicate module. The solution to this was to copy the fully_shard function and state and modify it for replicate. In the future, it should be explored making the replicate_state inherit from FSDP_state to get rid of code duplicity. I have attached below the profile tracing of a replicated Net Module.

https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/anshulsi_270fcc36-194a-42f5-9841-cace984c2132_devgpu263.prn2.facebook.com_1792146.1753232748025155780.pt.trace.json

**Test Case**
1.  pytest test/distributed/_composable/test_replicate_with_fsdp.py

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #158207

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta

### CODE REVIEW

File: test/distributed/_composable/test_replicate_with_fsdp.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+        replicate_model = deepcopy(model)
+
+        for i, layer in enumerate(replicate_model.layers):
+            if i % 2 == 0:
+                replicate(layer)
+            else:
+                fully_shard(layer)
+
+        # attempted to use parametrize before but it was not working
+        sharding_strategy = "shard"
```
Comment (mori360) [inline test/distributed/_composable/test_replicate_with_fsdp.py]: You can use run_subtest to test both sharding_strategy
Example here: https://github.com/pytorch/pytorch/blob/194539e9c33dc793fe67fbb68c7cee12f399e276/test/distributed/fsdp/test_fsdp_mixed_precision.py#L678-L684

File: torch/distributed/_composable/replicate_with_fsdp.py [hunk 2]
```diff
@@ -0,0 +1,679 @@
...
+    torch._C._log_api_usage_once("torch.distributed.fsdp.fully_shard")
+    if isinstance(module, (nn.ModuleList, nn.ModuleDict)):
+        raise ValueError(
+            f"replicate does not support containers that do not implement forward: {module}"
+        )
+    mesh = mesh or _init_default_fully_shard_mesh()
+    if mesh.ndim not in (1, 2):
+        raise ValueError(f"replicate expects a 1D or 2D DeviceMesh but got {mesh}")
+    elif mesh.ndim == 1:
+        mesh_info = FSDPMeshInfo(mesh, shard_mesh_dim=0)
```
Comment (mori360) [inline torch/distributed/_composable/replicate_with_fsdp.py]: Replicate does not accept FSDPMeshInfo here, and we will have future work to accept both mesh in ndim 1 or 2

File: torch/distributed/_composable/replicate_with_fsdp.py [hunk 3]
```diff
@@ -0,0 +1,679 @@
...
+            #
+            # This won't conflict with what is done in DDP class as the module
+            # replicate is going to pass is NOT the original module.
+            from torch.distributed.tensor.parallel.ddp import (
+                _localize_dtensor,
+                _reconstruct_dtensor,
+            )
+
+            module.register_forward_pre_hook(_reconstruct_dtensor)
+            module.register_forward_hook(_localize_dtensor)
```
Comment (mori360) [inline torch/distributed/_composable/replicate_with_fsdp.py:284]: Can we add a unit test to test device mesh is not None in kwargs.

File: test/distributed/_composable/test_replicate_with_fsdp.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+        model = Net()
+        replicate_model = deepcopy(model)
+
+        replicate(replicate_model.fc1)
+        fully_shard(replicate_model.fc2)
+
+        replicate_model = replicate(replicate_model)
+        managed_modules = _get_managed_modules((replicate_model,))
+
+        # self._compare_module(model, replicate_model)
```
Comment (mori360) [inline test/distributed/_composable/test_replicate_with_fsdp.py]: Can we add the module compare here?

File: torch/distributed/_composable/replicate_with_fsdp.py [hunk 5]
```diff
@@ -0,0 +1,679 @@
...
+    # Registry keys by function name
+    return "fully_shard" not in registry
+
+
+def _is_fully_sharded(module: nn.Module) -> bool:
+    r"""Check if module is marked with fully_shard."""
+    registry = _get_registry(module)
+    if registry is None:
+        return False
+    return "fully_shard" in registry
```
Comment (mori360) [inline torch/distributed/_composable/replicate_with_fsdp.py]: can we combine `is_composable_with_replicate` and `_is_fully_sharded`

File: test/distributed/_composable/test_replicate_with_fsdp.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+
+        for i, layer in enumerate(replicate_model.layers):
+            if i % 2 == 0:
+                self.assertTrue("replicate" in _get_registry(layer))
+                for parameter in layer.parameters():
+                    self.assertEqual(parameter.placements, (Replicate(), Shard(dim=0)))
+            else:
+                self.assertTrue("fully_shard" in _get_registry(layer))
+                for parameter in layer.parameters():
+                    self.assertEqual(parameter.placements, (Shard(dim=0),))
```
Comment (mori360) [inline test/distributed/_composable/test_replicate_with_fsdp.py]: Can we check the root's mesh here?

File: torch/distributed/_composable/replicate_with_fsdp.py (hunk 2 above)
Comment (anshul-si) [reply torch/distributed/_composable/replicate_with_fsdp.py]: Agreed with points above, will change to check if it doesn't equal 2 only. I think whenever we work on ndim 1, it will be changed before replicate_impl is called.

File: test/distributed/_composable/test_replicate_with_fsdp.py (hunk 4 above)
Comment (mori360) [reply test/distributed/_composable/test_replicate_with_fsdp.py]: Here's an example, https://github.com/pytorch/pytorch/blob/26807dcf277feb2d99ab88d7b6da526488baea93/test/distributed/_composable/fsdp/test_fully_shard_comm.py#L427-L456
Excluding divide_factor, we can compare the loss, and module as in line 455 and 456

File: test/distributed/_composable/test_replicate_with_fsdp.py [hunk 7]
```diff
@@ -88,6 +90,40 @@ def _init_pg(self):
             store=dist.FileStore(self.file_name, self.world_size),
         )
 
+    def check_sharded_parity(
```
Comment (mori360) [inline test/distributed/_composable/test_replicate_with_fsdp.py]: can we import it from fsdp_common?

File: test/distributed/_composable/test_replicate_with_fsdp.py [hunk 8]
```diff
@@ -120,15 +156,19 @@ def step_model(model, input, target):
...
-            for i, j in zip(mod.parameters(), replicate_mod.parameters()):
-                self.assertEqual(i, j, rtol=1.3e-06, atol=5e-5)
+            for x, y in zip(mod.parameters(), replicate_mod.parameters()):
+                self.assertEqual(x, y.to_local(), rtol=1.3e-06, atol=5e-5)
 
             # Shuffle the input so that DDP input is different
             torch.manual_seed(iteration)
             input = input[torch.randperm(global_batch_size)]
 
+    def run_subtests(self, *args, **kwargs):
```
Comment (mori360) [inline test/distributed/_composable/test_replicate_with_fsdp.py]: what's the difference between these 2 `run_subtests`

File: test/distributed/_composable/test_replicate_with_fsdp.py [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+            rank=self.rank,
+            world_size=self.world_size,
+            store=dist.FileStore(self.file_name, self.world_size),
+        )
+
+    def run_subtests(self, *args, **kwargs):
+        return run_subtests(self, *args, **kwargs)
+
+    @skip_if_lt_x_gpu(2)
+    def test_replicate_net_module(self):
```
Comment (mori360) [inline test/distributed/_composable/test_replicate_with_fsdp.py]: What's the difference between `test_replicate_transformer`, `test_replicate_managed_modules` and  `test_replicate_net_module`

File: test/distributed/_composable/test_replicate_with_fsdp.py [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+
+    def _init_pg(self):
+        dist.init_process_group(
+            backend="nccl",
+            rank=self.rank,
+            world_size=self.world_size,
+            store=dist.FileStore(self.file_name, self.world_size),
+        )
+
+    def _check_state_dict_parity(self, sd_1, sd_2):
```
Comment (mori360) [inline test/distributed/_composable/test_replicate_with_fsdp.py]: Is _check_state_dict_parity called somewhere?

File: test/distributed/_composable/test_replicate_with_fsdp.py [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+
+        for i, layer in enumerate(replicate_model.layers):
+            if i % 3 == 0:
+                self.assertTrue("replicate" in _get_registry(layer))
+                for parameter in layer.parameters():
+                    self.assertEqual(parameter.placements, (Replicate(), Shard(dim=0)))
+            elif i % 3 == 1:
+                self.assertTrue("fully_shard" in _get_registry(layer))
+                for parameter in layer.parameters():
+                    self.assertEqual(parameter.placements, (Shard(dim=0),))
```
Comment (mori360) [inline test/distributed/_composable/test_replicate_with_fsdp.py:127]: can we the check for the layers that i%3==2?

File: test/distributed/_composable/test_replicate_with_fsdp.py [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+        for v1, v2 in zip(sd_1.values(), sd_2.values()):
+            self.assertEqual(v1, v2)
+
+
+class ReplicateTest(MultiProcessTestCase):
+    @property
+    def world_size(self) -> int:
+        return 4
+
+    def init_global_mesh(self) -> DeviceMesh:
```
Comment (mori360) [inline test/distributed/_composable/test_replicate_with_fsdp.py]: can we rename here, since the test class is ReplicateTest, we need to point out the mesh here contains TP

File: test/distributed/_composable/test_replicate_with_fsdp.py [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+                for parameter in layer.parameters():
+                    self.assertEqual(parameter.placements, (Shard(dim=0),))
+
+        managed_modules = _get_managed_modules((replicate_model,))
+
+        if sharding_strategy == "replicate":
+            self.assertEqual(len(managed_modules), 7)
+
+    @skip_if_lt_x_gpu(2)
+    def test_replicate_device_mesh(self):
```
Comment (mori360) [inline test/distributed/_composable/test_replicate_with_fsdp.py]: rename here also.

File: test/distributed/_composable/test_replicate_with_fsdp.py [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+        layers = [
+            replicate_model.fc1,
+            replicate_model.fc2,
+            replicate_model.fc3,
+        ]
+
+        for layer in layers:
+            fully_shard(layer)
+
+        replicate_model = fully_shard(replicate_model)
```
Comment (mori360) [inline test/distributed/_composable/test_replicate_with_fsdp.py]: compare the model training with replicate

File: torch/distributed/_composable/replicate_with_fsdp.py [hunk 15]
```diff
@@ -0,0 +1,655 @@
...
+            dct = {"__deepcopy__": _unimplemented_deepcopy}
+            new_cls = type(f"FSDP{cls.__name__}", (FSDPModule, cls), dct)
+            cls_to_fsdp_cls[cls] = new_cls
+        module.__class__ = new_cls
+    return arg_module
+
+
+@contract(state_cls=_ReplicateState)
+def replicate(
+    module: nn.Module,
```
Comment (mori360) [inline torch/distributed/_composable/replicate_with_fsdp.py:238]: can we add the ignored_module back, we can utilize ignored_params in fully_shard to do the ignore later

File: torch/distributed/_composable/replicate_with_fsdp.py [hunk 16]
```diff
@@ -0,0 +1,655 @@
...
+        self._is_root: Optional[bool] = None  # root set during lazy init
+        self._state_ctx = _ReplicateStateContext()
+        self._comm_ctx = FSDPCommContext()
+        self._training_state: TrainingState = TrainingState.IDLE
+        self._states_to_forward_prefetch: list[FSDPState] = []
+        self._states_to_backward_prefetch: list[FSDPState] = []
+        self._modules_to_run_forward: set[nn.Module] = set()
+        # ``False`` when user set reshard_after_forward
+        # through ``fully_shard`` or ``set_reshard_after_forward``
+        self._auto_reshard_after_forward: Optional[bool] = True
```
Comment (mori360) [inline torch/distributed/_composable/replicate_with_fsdp.py]: TODO: discussion on whether to remove `self._auto_reshard_after_forward`

File: test/distributed/_composable/test_replicate_with_fsdp.py (hunk 11 above)
Comment (anshul-si) [reply test/distributed/_composable/test_replicate_with_fsdp.py:127]: I could check if they're still composable, but this is essentially being done in the managed_modules assertion

File: torch/distributed/_composable/replicate_with_fsdp.py [hunk 17]
```diff
@@ -0,0 +1,655 @@
...
+    ignored_params: Optional[set[nn.Parameter]] = None,
+):
+    torch._C._log_api_usage_once("torch.distributed.fsdp.fully_shard")
+    if isinstance(module, (nn.ModuleList, nn.ModuleDict)):
+        raise ValueError(
+            f"replicate does not support containers that do not implement forward: {module}"
+        )
+
+    mesh = mesh or _init_default_fully_shard_mesh()
+    if mesh.ndim != 2:
```
Comment (mori360) [inline torch/distributed/_composable/replicate_with_fsdp.py:183]: if user input 1d mesh(replicate mesh), we should init the 2d (replicate, shard) mesh and run replicate

File: test/distributed/_composable/test_replicate_with_fsdp.py [hunk 18]
```diff
@@ -0,0 +1,233 @@
...
+                    self.assertEqual(parameter.placements, (Replicate(), Shard(dim=0)))
+            elif i % 3 == 1:
+                self.assertTrue("fully_shard" in _get_registry(layer))
+                for parameter in layer.parameters():
+                    self.assertEqual(parameter.placements, (Shard(dim=0),))
+
+        managed_modules = _get_managed_modules((replicate_model,))
+
+        if sharding_strategy == "replicate":
+            self.assertEqual(len(managed_modules), 7)
```
Comment (mori360) [inline test/distributed/_composable/test_replicate_with_fsdp.py]: Where does 7 come from

File: test/distributed/_composable/test_replicate_with_fsdp.py (hunk 14 above)
Comment (mori360) [reply test/distributed/_composable/test_replicate_with_fsdp.py]: can we have another unit test to train with replicate and tp
https://github.com/pytorch/pytorch/blob/da05b7fb94fa6382c43e165a525a76d8ae62cadd/test/distributed/_composable/test_composability/test_2d_composability.py#L117-L137

File: torch/distributed/_composable/replicate_with_fsdp.py [hunk 19]
```diff
@@ -0,0 +1,655 @@
...
+def is_composable_with_replicate(module: nn.Module) -> bool:
+    """Checks if replicate can be applied with module"""
+    registry = _get_registry(module)
+    if registry is None:
+        return True
+    # Registry keys by function name
+    return "fully_shard" not in registry
+
+
+def replicate_mesh():
```
Comment (weifengpy) [inline torch/distributed/_composable/replicate_with_fsdp.py:338]: check if we create pg every time through profiler traces

File: torch/distributed/_composable/replicate_with_fsdp.py [hunk 20]
```diff
@@ -0,0 +1,399 @@
...
+
+if TYPE_CHECKING:
+    from torch.distributed.tensor import Shard
+
+
+cls_to_fsdp_cls: dict[type, type] = {}
+
+_ROOT_MODULE_PREFIX = ""
+
+logger = logging.getLogger("torch.distributed.fsdp.fully_shard")
```
Comment (weifengpy) [inline torch/distributed/_composable/replicate_with_fsdp.py]: logger = logging.getLogger("torch.distributed._composable.replicate_with_fsdp")

**Review by weifengpy — APPROVED** [review]
